wallet/src/de/schildbach/wallet/payments/SendCoinsTaskRunner.kt

```diff
@@ -0,0 +1,133 @@
+    ): Transaction = withContext(Dispatchers.IO) {
+        val securityGuard = SecurityGuard()
+        val password = securityGuard.retrievePassword()
+        val encryptionKey = deriveKey(wallet, password, scryptIterationsTarget)
+
+        sendRequest.aesKey = encryptionKey
+        sendRequest.exchangeRate = exchangeRate
+
+        try {
+            log.info("sending: {}", sendRequest)
+            Context.propagate(Constants.CONTEXT)
+
+            if (txCompleted) {
+                wallet.commitTx(sendRequest.tx)
+            } else {
+                wallet.sendCoinsOffline(sendRequest)
+            }
+
+            val transaction = sendRequest.tx
+            log.info("send successful, transaction committed: {}", transaction.txId.toString())
+            walletApplication.broadcastTransaction(transaction)
+            transaction
+        } catch (ex: Exception) {
+            when (ex) {
+                is InsufficientMoneyException -> ex.missing?.run {
+                    log.info("send failed, {} missing", toFriendlyString())
+                } ?: log.info("send failed, insufficient coins")
+                is ECKey.KeyIsEncryptedException -> log.info("send failed, key is encrypted: {}", ex.message)
+                is KeyCrypterException -> log.info("send failed, key crypter exception: {}", ex.message)
+                is Wallet.CouldNotAdjustDownwards -> log.info("send failed, could not adjust downwards: {}", ex.message)
+                is Wallet.CompletionException -> log.info("send failed, cannot complete: {}", ex.message)
+            }
+            throw ex
+        }
+    }
+
+    @Throws(KeyCrypterException::class)
+    private fun deriveKey(wallet: Wallet, password: String, scryptIterationsTarget: Int): KeyParameter {
+        Preconditions.checkState(wallet.isEncrypted)
+        val keyCrypter = wallet.keyCrypter!!
+
+        // Key derivation takes time.
+        var key = keyCrypter.deriveKey(password)
+
+        // If the key isn't derived using the desired parameters, derive a new key.
+        if (keyCrypter is KeyCrypterScrypt) {
+            val scryptIterations = keyCrypter.scryptParameters.n
+
+            if (scryptIterations != scryptIterationsTarget.toLong()) {
+                log.info("upgrading scrypt iterations from {} to {}; re-encrypting wallet",
+                        scryptIterations, scryptIterationsTarget)
+                val newKeyCrypter = KeyCrypterScrypt(scryptIterationsTarget)
+                val newKey: KeyParameter = newKeyCrypter.deriveKey(password)
+
+                // Re-encrypt wallet with new key.
+                try {
+                    wallet.changeEncryptionKey(newKeyCrypter, key, newKey)
+                    key = newKey
+                    log.info("scrypt upgrade succeeded")
+                } catch (x: KeyCrypterException) {
+                    log.info("scrypt upgrade failed: {}", x.message)
+                }
+            }
+        }
```

Comment:
I was wondering why this was here, but it does come from `DeriveKeyTask`

In the parent app, Bitcoin Wallet, the number of scrypt hash function iterations has increased over time.  For Krip Phones, the number of iterations is set to a smaller number than normal.  If we ever increase that number in a future version or if a user takes a backup file made from Krip Phone or a previous version with a lower number of iterations, then this code would change the key that encrypts the wallet.